subprojects/core/src/integTest/groovy/org/gradle/caching/internal/packaging/BuildCacheEntryPackingIntegrationTest.groovy

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.caching.internal.packaging
+
+import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+import org.gradle.integtests.fixtures.daemon.DaemonIntegrationSpec
+import spock.lang.Issue
+import spock.lang.Unroll
+
+class BuildCacheEntryPackingIntegrationTest extends DaemonIntegrationSpec implements DirectoryBuildCacheFixture {
+
+    private static final NON_ASCII_NAME = [
+        "ascii-only": "ascii",
+        "space": " ",
+        "zwnj": "\u200c",
+        "chinese": "敏捷的棕色狐狸跳过了懒狗",
+        "cyrillic": "здравствуйте",
+        "hungarian": "Árvíztűrő tükörfúrógép",
+    ].values().join("-")
+
+    private static final DEFAULT_ENCODINGS = [
+        "UTF-8",
+        "ISO-8859-1",
+        "windows-1250",
+    ]
+
+    @Unroll
+    def "can store and load files having non-ascii characters in file name using default file encoding #fileEncoding"() {
+        def fileName = NON_ASCII_NAME + ".txt"
+        def outputFile = file("dir", fileName)
+
+        buildFile << """
+            println "> Default charset: \${java.nio.charset.Charset.defaultCharset()}"
+            println "> Storing file in cache: $fileName"
+
+            task createFile {
+                outputs.dir("dir")
+                outputs.cacheIf { true }
+                doLast {
+                    file("dir/$fileName").text = "output"
+                }
+            }
+        """
+
+        when:
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        executedAndNotSkipped(":createFile")
+
+        when:
+        assert outputFile.delete()
+        withBuildCache().run("createFile", "-Dfile.encoding=$fileEncoding")
+        skipped(":createFile")
+
+        then:
+        output.contains("> Default charset: $fileEncoding")
+        outputFile.text == "output"
+
+        where:
+        fileEncoding << DEFAULT_ENCODINGS
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/9877")
+    @Unroll
+    def "can store and load files having non-ascii characters in property name using default file encoding #fileEncoding"() {
```

Comment:
❓ are we actually using the default encoding or rather ignoring it?

Reply:
We are using a default encoding to run the build.

Reply:
Didn't we specify UTF-8 explicitly to pack/unpack - or is that not what the title refers to?

Reply:
Yes, that is how we deal with the default encoding (in the OS) being set to different values.

Reply:
I understood the encoding is ignored in favor of the encoding we're explicitly setting (utf8). Do I misunderstand the title or the fix?

Reply:
I'm not sure what you are asking. Do you have a suggestion for a change maybe?

Reply:
Before setting the encoding explicitly, the default encoding was used, right?
So the test name does reflect the previous state, but now we're explicitly setting the encoding that should be used, so it seems to me that the default file encoding (still present in the test name) doesn't play any role anymore.
I'm okay with you ignoring this nit, I may have misunderstood it.

Reply:
We don't set the default encoding/charset set for the JVM. What we do in production code (and was not doing well before the fix) is that we keep our business independent of the default encoding. This is what the test title says: (even) when default encoding is set to something unusual we succeed in conducting our work. Let me know if you have an idea how to clarify this further if needed.

Reply:
Yes, that's why maybe instead we could:
> can store and load files [...] ~using~ ignoring the default file encoding

If you don't think that helps, please feel free to resolve this conversation.

Reply:
OK, I pushed a change, please check.